src/net/ifaddrs/mod.rs

```diff
@@ -0,0 +1,273 @@
+//! more convenient access by interface name.
+//!
+//! # Examples
+//!
+//! You can access the basic information of the system in a few lines.
+//! The following program prints all the known addresses.
+//!
+//! ```
+//! use nix::net::ifaddrs::InterfaceAddrs;
+//!
+//! let addrs = InterfaceAddrs::getifaddrs()
+//!     .expect("Failed to enumerate network interfaces.");
+//!
+//! for addr in addrs {
+//!     println!("{}: {:?}", addr.name, addr.address);
+//! }
+//! ```
+//!
+//! The `IffFlags` struct provides access to info about the
+//! state of an interface. This program prints the addresses of only
+//! interfaces which are up.
+//!
+//! ```
+//! use nix::net::ifaddrs::{InterfaceAddrs, iff_flags};
+//!
+//! let addrs = InterfaceAddrs::getifaddrs()
+//!     .expect("Failed to eunmerate network interfaces.");
+//!
+//! for addr in addrs {
+//!     if addr.flags.contains(iff_flags::IFF_UP) {
+//!         println!("{}: {:?}", addr.name, addr.address);
+//!     }
+//! }
+//! ```
+//!
+//! You can convert the `InterfaceAddrs` struct into a `HashMap` easily.
+//! `InterfaceMap` is an alias for `HashMap<String, Vec<InterfaceAddr>>` for
+//! easier reference.
+//!
+//! ```
+//! use nix::net::ifaddrs::{InterfaceAddrs, InterfaceAddr, InterfaceMap};
+//! use std::collections::HashMap;
+//!
+//! let interfaces: InterfaceMap =
+//!     InterfaceAddrs::getifaddrs()
+//!     .expect("Failed to enumerate network interfaces.")
+//!     .into(); // Convert to a hash map
+//!
+//! // Print all the addresses of the loopback interface
+//! if let Some(addrs) = interfaces.get("lo") {
+//!    println!("Loopback addresses:");
+//!    for addr in addrs {
+//!         println!("\t{:?}", addr);
+//!    }
+//! }
+//!
+//! ```
+//!
+
+use libc;
+use std::ptr::null_mut;
+use std::ffi::CStr;
+use std::collections::HashMap;
+use errno::{Errno, errno};
+use Error;
+use Result;
+
+pub mod iff_flags;
+use self::iff_flags::IffFlags;
+
+mod sockaddr;
+use self::sockaddr::{IfAddrValue, sockaddr_to_ifaddrvalue};
+
+mod tests;
+
+pub type InterfaceMap<'a> = HashMap<String, Vec<InterfaceAddr<'a>>>;
+
+/// Represents a handle into the operating system's knowledge about network
+/// interfaces present on the system. Allows the user to iterate over
+/// interface configurations.
+pub struct InterfaceAddrs<'a> {
+    inner: *mut libc::ifaddrs,
+    current: Option<&'a libc::ifaddrs>,
+    do_free: bool,
+}
+
+impl<'a> InterfaceAddrs<'a> {
+    /// Creates an `InterfaceAddrs` from a raw pointer, without calling into
+    /// the `libc`.
+    ///
+    /// The destructor will not attempt to free memory on an InterfaceAddrs
+    /// created in this way.
+    ///
+    /// # Unsafety
+    /// The caller is responsible for making sure the given pointer is not
+    /// in invalid memory.
+    ///
+    /// # Errors
+    /// `Err(())` will be returned if `p` was void.
+    pub unsafe fn from_raw(p: *mut libc::ifaddrs) -> ::std::result::Result<InterfaceAddrs<'a>, ()> {
```

Comment:
I'm trying to understand the use case for having this in our API? Is this really necessary for the user to do everything they want to do with `ifaddrs`? Where would the user get a struct like this rather than just have our own home-grown constructor that hides the struct creation from the user?